From 77e3d0b95d708542afa3dd3845b841268a2e0b80 Mon Sep 17 00:00:00 2001
From: Konrad Rzeszutek Wilk <konrad.wilk@oracle.com>
Date: Mon, 29 Nov 2010 14:03:30 -0500
Subject: [PATCH] ttm: Utilize the dma_addr_t array for pages that are to in DMA32 pool.

We only use the "if (pool == NULL)" path for right now.

Based on fcba97cde2f from
git://git.kernel.org/pub/scm/linux/kernel/git/konrad/xen.git#devel/ttm.pci-api-v2
by Konrad Rzeszutek Wilk <konrad.wilk@oracle.com> but completely rewritten due
to the absence of 1403b1a38e8b19a "drm/ttm: add pool wc/uc page allocator V3"
in this tree.

Signed-off-by: Ian Campbell <ian.campbell@citrix.com>
---
 drivers/gpu/drm/ttm/ttm_tt.c |   23 +++++++++++++++++++++--
 1 files changed, 21 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/ttm/ttm_tt.c b/drivers/gpu/drm/ttm/ttm_tt.c
index 4ff26c6..dc257bf 100644
--- a/drivers/gpu/drm/ttm/ttm_tt.c
+++ b/drivers/gpu/drm/ttm/ttm_tt.c
@@ -34,6 +34,7 @@
 #include <linux/pagemap.h>
 #include <linux/file.h>
 #include <linux/swap.h>
+#include <linux/dma-mapping.h>
 #include "drm_cache.h"
 #include "ttm/ttm_module.h"
 #include "ttm/ttm_bo_driver.h"
@@ -102,12 +103,30 @@ static struct page *ttm_tt_alloc_page(unsigned page_flags, dma_addr_t *dma_addre
 	else
 		gfp_flags |= __GFP_HIGHMEM;
 
-	return alloc_page(gfp_flags);
+	if ((page_flags & TTM_PAGE_FLAG_DMA32) && dma_address) {
+		void *addr;
+		addr = dma_alloc_coherent(NULL, PAGE_SIZE,
+					  dma_address,
+					  gfp_flags);
+		if (addr == NULL)
+			return NULL;
+		return virt_to_page(addr);
+	} else
+		return alloc_page(gfp_flags);
 }
 
 static void ttm_tt_free_page(struct page *page, unsigned page_flags, dma_addr_t *dma_address)
 {
-	__free_page(page);
+	if ((page_flags & TTM_PAGE_FLAG_DMA32) && dma_address) {
+		void *addr = page_address(page);
+		WARN_ON(!addr || !(*dma_address));
+		if (addr)
+			dma_free_coherent(NULL, PAGE_SIZE,
+					 addr,
+					 *dma_address);
+		*dma_address = 0;
+	} else
+		__free_page(page);
 }
 
 static void ttm_tt_free_user_pages(struct ttm_tt *ttm)
-- 
1.5.6.5

